File: waveCat.py
import os
import numpy as np
import scipy.signal 
from scipy.io.wavfile import read
from scipy.io.wavfile import write
import re

class TextToSpeech():
    def __init__(self):
        self.load_words('cmudict-0.7b.txt')

    def containsVowel(self,instr):
        #0 if no vowel, 1 if starts with vowel, 2 if ends with vowel, 3 if both
        #instr is a list
        vowels = 'aeiouAEIOU'
        if instr[0] in vowels and instr[-1] in vowels:
            result = 3
        elif instr[-1] in vowels:
            result = 2
        elif instr[0] in vowels:
            result = 1
        else:
            result = 0
        return result

    def vowelClip(self, phon, phonetic):
        cliplen = np.floor(len(phon)*0.2)
        result = self.containsVowel(phonetic)
        out = phon
        # if result == 2:
        #     out = phon[cliplen:]
        # elif result == 1:
        #     cliplenhalf = np.floor(cliplen*0.5)
        #     out = phon[:len(phon)-cliplenhalf]
        if result == 0:
            out = phon[0:int(len(phon)-cliplen)]

        return out

    def wavClip(self, inStream):
        maxval = max(inStream)
        threshold = 0.2
        inFloor = np.zeros(len(inStream))
        inFloor[:] = [1 if ele > threshold*maxval else 0 for ele in inStream]
        idx = np.flatnonzero(inFloor)
        idxStart = idx[0]
        idxEnd = idx[-1]
        if len(idxStart) == 0:
            idxStart = 1
        if len(idxEnd) == 0:
            idxEnd = len(inStream)

        idxStart = idxStart[0]
        idxEnd = idxEnd[-1]
        outStream = inStream[idxStart:idxEnd]
        return outStream


    def load_words(self, words_pron_dict:str):
        self.arpabet = {}
        with open(words_pron_dict, 'r') as file:
            for line in file:
                if not line.startswith(';;;'):
                    key, val = line.split('  ',2)
                    self.arpabet[key] = re.findall(r"[A-Z]+",val)
        return self.arpabet


    def get_pronunciation(self, str_input, filename):
        # Perform actual concatenation of input phonetics
        #arpabet = load_words('cmudict-0.7b.txt')
        list_pron = []
        completesentence = []
        print(str_input)
        print(str_input.split())
        for item in str_input.split():
            #if '?' or ',' or '.' or '!' in item[-1]:
            if item[-1] == '?' or item[-1] == '!' or item[-1] == '.' or item[-1] == ',':
                completesentence.append(item[:-1])
                completesentence.append(item[-1])
                completesentence.append('_') #input space
            elif item[-1] == ';' or item[-1] == ':' or item[-1] == '~':
                completesentence.append(item[:-1])
                completesentence.append('_')
            else:
                completesentence.append(item)
                completesentence.append('_')
        completesentence = completesentence[:-1]
        print(completesentence)
        # completesentence = re.findall(r"[\w']+",str_input)
        # #insert blank space between words in list
        # i=0
        # while i < len(completesentence):
        #     completesentence = completesentence[:i] + ["ZZZ"] + completesentence[i:]
        #     i += 2
        # print(completesentence[1:])      
        for word in completesentence:
            if word == '?':
                word = 'QNM'
            elif word == '!':
                word = 'EXM'
            elif word == ',':
                word = 'CMA'
            elif word == '.':
                word = 'FST'
            elif word == '_':
                word = 'SPC'
            if word.upper() in self.arpabet: #if the word is in the dictionary
                list_pron += self.arpabet[word.upper()]
        print(list_pron)
        testSentence = list_pron
        #testSentence = ['DH', 'AH', 'SPC', 'K', 'W', 'IH', 'K', 'SPC', 'B', 'R', 'AW', 'N', 'SPC', 'F', 'AA', 'K', 'S', 'CMA', 'SPC', 'JH', 'AH', 'M', 'P', 'T', 'SPC', 'OW', 'V', 'ER', 'SPC', 'DH', 'AH', 'SPC', 'W', 'AO', 'L', 'FST', 'SPC', 'S', 'AE', 'L', 'IY', 'SPC', 'W', 'AA', 'Z', 'SPC', 'N', 'AA', 'T', 'SPC', 'P', 'L', 'IY', 'Z', 'D', 'FST']
        testSentence = [item.lower() for item in testSentence]
        datastream = np.ndarray([0],dtype=np.float32);
        # joints = [] #For method 2
        joints = np.ndarray([0])
        Fs = 44100

        #parsing spaces and commas and other punctuations
        for i in range(len(testSentence)):
            item = testSentence[i]
            if item == 'spc':
                phon = np.zeros(int(Fs*0.1))
            elif item == 'cma':
                phon = np.zeros(int(Fs*0.3))
            elif item == 'fst':
                phon = np.zeros(int(Fs*0.4))
            elif item == 'exm':
                phon = np.zeros(int(Fs*0.4))
            elif item == 'qnm':
                phon = np.zeros(int(Fs*0.4))
            else:
                file = filename + testSentence[i] + ".wav"
                print (file)
                Fs, phon = scipy.io.wavfile.read(file)
                phon = self.vowelClip(phon, item)
            print (phon)
            datastream = np.append(datastream, phon);
            if i < len(testSentence):
                joints = np.append(joints, len(datastream))
        scipy.io.wavfile.write("phon.wav", Fs, phon)
        #Apply low pass filtering to smoothen the voice
        b,a = scipy.signal.butter(4,1500/(Fs/2))
        dataout = scipy.signal.lfilter(b,a,datastream)
        tmp = np.ndarray([0],dtype=np.float32)

        tmp = dataout
        tmp = tmp.astype(np.float32)

        #write audio to file

        tmp = tmp / np.max(tmp)
        scipy.io.wavfile.write('messages/datastream.wav', Fs, tmp)

if __name__ == '__main__':
    tts = TextToSpeech()
    tts.get_pronunciation('hey guys, i sound like you.','clippedAudio/ashley/')

File: waveCat.m
clear;
tic;
lookupfolder = 'clippedAudio';
destfolder = 'messages';
destfilename = 'matlabaudio';

% % Clean audio clips containing individual phonetics
<<<<<<< HEAD
f = dir('damon/*.wav');
for i = 1:length(f)
    currfile = f(i);
    [audioData, Fs] = audioread(sprintf('%s//%s',currfile.folder,currfile.name));
    clipped = wavClip(audioData);
    audiowrite(sprintf('%s//%s',lookupfolder, currfile.name), clipped, Fs);
end
=======
% f = dir('ger_eg/*.wav');
% for i = 1:length(f)
%     currfile = f(i);
%     [audioData, Fs] = audioread(sprintf('%s//%s',currfile.folder,currfile.name));
%     clipped = wavClip(audioData);
%     audiowrite(sprintf('%s//%s',lookupfolder, currfile.name), clipped, Fs);
% end
>>>>>>> 8fbf7c8ed016d386dd6b9f3656973bc4cc85a5bb

% Perform actual concatenation of input phonetics
% testSentence = {'HH', 'AH', 'L', 'OW', 'AY', 'AE', 'M', 'AH', 'R', 'OW', 'B', 'AA', 'T', 'AY', 'L', 'AY', 'K', 'T', 'UW', 'IY', 'T', 'AE', 'P', 'AH', 'L', 'Z'};
testSentence = {'HH', 'AH', 'L', 'OW',...
    ',',...
    'AY',...
    ' ',...
    'AE', 'M',...
    ' ',...
    'AH',...
    ' ',...
    'R', 'OW', 'B', 'AA', 'T',...
    '.',...
    'AY',...
    ' ',...
    'L', 'AY', 'K',...
    ' ',...
    'T', 'UW',...
    ' ',...
    'IY', 'T',...
    ' ',...
    'AE', 'P', 'AH', 'L', 'Z'};
<<<<<<< HEAD
testSentence = {'DH', 'EH', 'R', 'Z', 'ZZZ', 'AH', 'ZZZ', 'HH', 'AH', 'N', 'D', 'R', 'AH', 'D', 'ZZZ', 'AH', 'N', 'D', 'ZZZ', 'F', 'AO', 'R', 'ZZZ', 'D', 'EY', 'Z', 'ZZZ', 'AH', 'V', 'ZZZ', 'S', 'AH', 'M', 'ER', 'ZZZ', 'V', 'EY', 'K', 'EY', 'SH', 'AH', 'N', 'ZZZ', 'DH', 'EH', 'N', 'ZZZ', 'S', 'K', 'UW', 'L', 'ZZZ', 'K', 'AH', 'M', 'Z', 'ZZZ', 'AH', 'L', 'AO', 'NG', 'ZZZ', 'JH', 'AH', 'S', 'T', 'ZZZ', 'T', 'UW', 'ZZZ', 'EH', 'N', 'D', 'ZZZ', 'IH', 'T'};

testSentence = lower(testSentence);

datastream = [];
=======
% testSentence = {'DH', 'EH', 'R', 'Z', 'ZZZ', 'AH', 'ZZZ', 'HH', 'AH', 'N', 'D', 'R', 'AH', 'D', 'ZZZ', 'AH', 'N', 'D', 'ZZZ', 'F', 'AO', 'R', 'ZZZ', 'D', 'EY', 'Z', 'ZZZ', 'AH', 'V', 'ZZZ', 'S', 'AH', 'M', 'ER', 'ZZZ', 'V', 'EY', 'K', 'EY', 'SH', 'AH', 'N', 'ZZZ', 'DH', 'EH', 'N', 'ZZZ', 'S', 'K', 'UW', 'L', 'ZZZ', 'K', 'AH', 'M', 'Z', 'ZZZ', 'AH', 'L', 'AO', 'NG', 'ZZZ', 'JH', 'AH', 'S', 'T', 'ZZZ', 'T', 'UW', 'ZZZ', 'EH', 'N', 'D', 'ZZZ', 'IH', 'T'};

testSentence = lower(testSentence);

datastream = []; 
>>>>>>> 8fbf7c8ed016d386dd6b9f3656973bc4cc85a5bb
joints = []; % For method 2
Fs = 44100;
for i = 1:length(testSentence)
    if contains(testSentence{i},' ') || contains(testSentence{i},'zzz')
        phon = zeros(Fs*0.1, 1);
    elseif testSentence{i} == ','
        phon = zeros(Fs*0.2, 1);
    elseif testSentence{i} == '.'
        phon = zeros(Fs*0.3, 1);
    else
<<<<<<< HEAD
        [phon, Fs] = audioread(sprintf('%s//%s.wav',lookupfolder,testSentence{i}));
%         [phon, Fs] = audioread(sprintf('%s//%s.wav',lookupfolder,testSentence{i}), 'native');
%         phon = double(phon);
=======
        [phon, Fs] = audioread(sprintf('%s//%s.wav',lookupfolder,testSentence{i}), 'native');
>>>>>>> 8fbf7c8ed016d386dd6b9f3656973bc4cc85a5bb
        phon = vowelClip(phon, testSentence{i});
    end
    datastream = [datastream, phon']; 
    if i < length(testSentence)
        joints = [joints, length(datastream)];
    end
end
%   Apply some filtering to smoothen the voice   %
% #############   TRY THINGS HERE   ############################
dataout = datastream;
% % % %   Method 1: simple moving average filter
% % % N = 10;
% % % dataout = filter(ones(1,N)./N,1,datastream);

%   Method 2: Minimize joining difference
<<<<<<< HEAD
% dataout = easyJoin(datastream, joints, Fs);

%   Method 4: easyJoin V2 - interpolation
dataout = datastream;
ninterp = Fs*0.1;
if joints(1) < ninterp
    ninterp = joints(1)-2;
end
if length(dataout)-joints(end) < ninterp
    ninterp = length(dataout)-joints(end)-2;
end
for i = 1:length(joints)
    before = dataout((joints(i)-ninterp-1):(joints(i)-1));
    after = dataout((joints(i)):(joints(i)+ninterp));
    combined = easyInterp(before,after,ninterp-2);
    dataout((joints(i)-ninterp-1):(joints(i)+ninterp)) = combined(1:(end-1));
end
=======
dataout = easyJoin(datastream, joints, Fs);
>>>>>>> 8fbf7c8ed016d386dd6b9f3656973bc4cc85a5bb

%   Method 3: Low Pass Filter to remove blips
[b,a] = butter(4, 1500/(Fs/2));
dataout = filter(b,a,dataout);

audiowrite(sprintf('%s//%s.wav',destfolder,destfilename), dataout, Fs);
toc;
sound(audioread(sprintf('%s//%s.wav',destfolder,destfilename)), Fs);







